heartbeat/updater.go

```diff
@@ -0,0 +1,53 @@
+package heartbeat
+
+import "time"
+
+// An Updater represents a function called at each Heart's update interval. It
+// is responsible for preforming any action the heart should preform at the
+// aforementioned interval.
+//
+// If an error is encountered while processing, it may be returned, and the
+// execution of the Updater may be halted.
+type Updater func(h *Heart, now time.Time) error
+
+const (
+	// ISO8601TimeFormat is a time format string according to the time
+	// package and is used to marshal time.Time instances into
+	// ISO8601-compliant strings.
+	//
+	// See: https://www.ietf.org/rfc/rfc3339.txt for more details.
+	ISO8601TimeFormat string = "2006-01-02T15:04:05"
+)
+
+var (
+	// HashUpdateTicker is a default implementation of the Updater type. It
+	// assumes the HASH type is used in Redis to store heartbeat updates
+	// with the following structure:
+	//
+	// ```
+	// 0) ...
+	// 1) HEARTBEAT_ID -> LAST_UPDATE_TIME
+	// 2) ...
+	// 3) ...
+	// ```
+	//
+	// The contents of the HASH are stored at the Heart's Keyspace(), and
+	// the individual HEARTBEAT_IDs are stored at the Heart's Name().
+	//
+	// The `LAST_UPDATE_TIME` is stored as an ISO8601 DateTime, according
+	// to the `ISO8601TimeFormat` constant above.
+	HashUpdateTicker Updater = func(h *Heart, now time.Time) error {
+		redis := h.Pool().Get()
+		defer redis.Close()
+
+		formatted := now.Format(ISO8601TimeFormat)
+
+		if _, err := redis.Do("HSET", h.Location().Keyspace(),
+			h.Location().Name(), formatted); err != nil {
+
+			return err
+		}
+
+		return nil
+	}
```

Comment:
Now, if I wanted to have heartbeats and check then, I'd have to do something like the following:

```
l := keys.NewLocation(":", "heartbeats")
heartbeat.New(heartbeat.HeartbeatParam{
    Pool: Pool,
    Location: keys.NewLocation(":", "heartbeats"),
    Interval: time.Second,
    Updater: heartbeat.HashUpdateTicker,
})
// ...
heartbeat.NewHashDetector(heartbeat.HashDetectorParam{
    Pool: Pool,
    Location: keys.NewLocation(":", "heartbeats"),
    Interval: time.Second * 2,
})
```

You (the consumer) is passing in the same stuff twice. You have to make _sure_ that stuff all matches and that you're using the right detector for the updater -- there's no constraints for this, and if you accidentally mess it up it there might not ever be an error, your stuff will just start failing silently. You also have to read all the documentation for all of these parts separately to figure out how they go together, there's not a unified interface to access any of it.

Contrast this with something like Redigo itself. It contains lots of components -- like pubsub, pooling, etc., but it's accessed from a unified API. You read some of the info for this API and your set, you don't have to read about any parts that might not be relevant to what you're doing. It's all self-contained and top-down, so you configure it once and you're set.